### PR TITLE
Add get_random_file_from_dir to gears.filesystem

### PR DESCRIPTION
**Pull request by Sorky**
Get the name of a random file from a given directory

### CODE REVIEW

**Comment by Sorky** [comment]
Derived from https://github.com/awesomeWM/awesome/issues/2596

**Comment by Elv13** [comment]
from the CI

    lib/gears/filesystem.lua:167:22: unused loop variable i

please use `_` for the name of unused variables to prevent the linter from having this warning

File: lib/gears/filesystem.lua [hunk 1]
```diff
@@ -153,4 +153,30 @@ function filesystem.get_dir(d)
     end
 end
 
+math.randomseed( os.clock() % 1 * 1e6 )
+--- Get the name of a random file from a given directory.
```
Comment (Elv13) [inline lib/gears/filesystem.lua]: Please add a newline before the `---`

File: lib/gears/filesystem.lua [hunk 2]
```diff
@@ -153,4 +153,30 @@ function filesystem.get_dir(d)
...
+-- @tparam string path The directory to search.
+-- @tparam[opt] table exts Specific extensions [eg: { "jpg", "png" }] to limit
```
Comment (Elv13) [inline lib/gears/filesystem.lua]: please change to \`{ "jpg", "png" }\`

File: lib/gears/filesystem.lua (hunk 1 above)
Comment (Sorky) [reply lib/gears/filesystem.lua]: Will do

File: lib/gears/filesystem.lua (hunk 2 above)
Comment (Sorky) [reply lib/gears/filesystem.lua]: This suit?
```
-- @tparam[opt] table exts Specific extensions to limit the search to.
--   If ommited, all files are considered. eg:`{ "jpg", "png" }`
```

**Comment by Sorky** [comment]
> Beside that a little unit or integration test would be welcome to make our coverage bot happy. You can't prove randomness, but at least our CI can hit the code and may help detect if, for example, a LGI change was to break it.

Would you like a spec/gears/filesystem_spec.lua

**Comment by Elv13** [comment]
> Would you like a spec/gears/filesystem_spec.lua?

This would be appreciated yes.

File: lib/gears/filesystem.lua (hunk 2 above)
Comment (Elv13) [reply lib/gears/filesystem.lua]: yes, thank you

Comment (Sorky) [reply lib/gears/filesystem.lua]: Updating to...
```
-- @tparam[opt] table exts Specific extensions to limit the search to. eg:`{ "jpg", "png" }`
--   If ommited, all files are considered.
```

**Comment by Sorky** [comment]
Added test for my changes plus several other existing areas.

Extra tests for areas outside of the scope of this PR that could be added marked as `-- TODO:`

File: lib/gears/filesystem.lua [hunk 3]
```diff
@@ -153,4 +153,33 @@ function filesystem.get_dir(d)
...
+        if file:get_file_type() == "REGULAR" then
+            local file_name = file:get_display_name()
+            if not exts or valid_exts[file_name:lower():match("%.(.*)$")] then
+               table.insert(files, file_name)
+            end
+        end
+    end
+
+    -- Return a randomly selected filename from the file table
+    return #files > 0 and files[math.random(#files)] or ""
```
[the remark this answers is not in the document]
Comment (Sorky) [reply lib/gears/filesystem.lua]: OK

**Comment by Elv13** [comment]
The tests are failing in the latest revision. Once fixed, I will approve the pull request, the code itself is ready to merge as far as I am concerned.

**Comment by Sorky** [comment]
> [#2596 (comment)](https://github.com/awesomeWM/awesome/issues/2596#issuecomment-459641931)
> 
> It was in my category of "to be handled once someone complains" because too many review cycle suck. But... Someone complained.

A problem is a problem - No need to wait - God forbid I found one in my tests which I improved as well

**Comment by Veratil** [comment]
Just a side thought which would be a refactor of the code, but it could use gears.string.endswith?

**Comment by Elv13** [comment]
> Just a side thought which would be a refactor of the code, but it could use gears.string.endswith?

That would make supporting multiple extensions a bit problematic, right? Given some people with have the bright idea of downloading a "The best 1 million wallpapers for 2019" style packs, looping string comparison instead of compiled pattern is a potential issue.

**Comment by Veratil** [comment]
> > Just a side thought which would be a refactor of the code, but it could use gears.string.endswith?
> 
> That would make supporting multiple extensions a bit problematic, right? Given some people with have the bright idea of downloading a "The best 1 million wallpapers for 2019" style packs, looping string comparison instead of compiled pattern is a potential issue.

True true. It would probably be looping through and checking each extension. Like I said just a side thought. 😉

**Comment by Sorky** [comment]
Updated regex, added all "." in filename tests and ensured { "" } works as well

Ready for the next round

**Comment by Sorky** [comment]
Not sure how the timing is going but as this is new and heavily tested and reviewed it might be good for 4.3?

**Comment by Elv13** [comment]
> Not sure how the timing is going but as this is new and heavily tested and reviewed it might be good for 4.3?

4.3 is already released, so this will get into v4.4. Distributions don't backport features, so even *if* I stop being against bugfixes releases, it wouldn't get integrated into existing packaged versions anyway. Right now I prefer to only release X.y versions because in the past we had X.y.z versions and git-master ended up thousands of commits ahead of the stable branch and ended up never being released.

As for this PR, well, time's up, I was waiting for the second review to auto-merge it, but this isn't happening, so I merge it.

**Comment by Sorky** [comment]
Any idea when the 4.3 release will get to the ubuntu apt?

**Comment by Elv13** [comment]
In the current Ubuntu? Never.

It will be in the next upgrade, so 19.10 or 20.04 depending if you use LTS or not. They don't seem interested in pulling it for 19.04, maybe nobody asked for it.

It's not a big deal, v4.2 was a solid release.

**Pull request merged**